storage/operations.go

```diff
@@ -0,0 +1,184 @@
+
+func DefaultIteratorOptions() IteratorOption {
+	return IteratorOption{
+		IterateKeyOnly: false, // only needed for badger. ignored by pebble
+	}
+}
+
+type Reader interface {
+	// Get gets the value for the given key. It returns ErrNotFound if the DB
+	// does not contain the key.
+	// other errors are exceptions
+	//
+	// The caller should not modify the contents of the returned slice, but it is
+	// safe to modify the contents of the `key` argument after Get returns. The
+	// returned slice will remain valid until the returned Closer is closed.
+	// when err == nil, the caller MUST call closer.Close() or a memory leak will occur.
+	Get(key []byte) (value []byte, closer io.Closer, err error)
+
+	// NewIter returns a new Iterator for the given key prefix range [startPrefix, endPrefix], both inclusive.
+	// We require that startPrefix ≤ endPrefix (otherwise this function errors).
+	// Specifically, all keys that meet ANY of the following conditions are included in the iteration:
+	//   - have a prefix equal to startPrefix OR
+	//   - have a prefix equal to the endPrefix OR
+	//   - have a prefix that is lexicographically between startPrefix and endPrefix
+	NewIter(startPrefix, endPrefix []byte, ops IteratorOption) (Iterator, error)
+}
+
+// Writer is an interface for batch writing to a storage backend.
+// One Writer instance cannot be used concurrently by multiple goroutines.
+type Writer interface {
+	// Set sets the value for the given key. It overwrites any previous value
+	// for that key; a DB is not a multi-map.
+	//
+	// It is safe to modify the contents of the arguments after Set returns.
+	// No errors expected during normal operation
+	Set(k, v []byte) error
+
+	// Delete deletes the value for the given key. Deletes are blind all will
+	// succeed even if the given key does not exist.
+	//
+	// It is safe to modify the contents of the arguments after Delete returns.
+	// No errors expected during normal operation
+	Delete(key []byte) error
+
+	// DeleteByRange removes all keys with a prefix that falls within the
+	// range [start, end], both inclusive.
+	// No errors expected during normal operation
+	DeleteByRange(globalReader Reader, startPrefix, endPrefix []byte) error
+}
+
+// ReaderBatchWriter is an interface for reading and writing to a storage backend.
+// It is useful for performing a related sequence of reads and writes, after which you would like
+// to modify some non-database state if the sequence completed successfully (via AddCallback).
+// If you are not using AddCallback, avoid using ReaderBatchWriter: use Reader and Writer directly.
+type ReaderBatchWriter interface {
+	// GlobalReader returns a database-backed reader which reads the latest committed global database state ("read-committed isolation").
+	// This reader will not read writes written to ReaderBatchWriter.Writer until the write batch is committed.
+	// This reader may observe different values for the same key on subsequent reads.
+	GlobalReader() Reader
```

Comment:
Do you mean why the GlobalReader is part of the Batch? 

It allows us to read data and depending the result write different data to the batch. The Reader is part of the Writer so that it's ensure the underlying database instance is the same. And the reason its called GlobalReader is to indicate its reading the committed state.

Reply:
I meant to place the comment at the name of the interface.

What I was trying to say is that the interface name (or purpose) is just to be a `Batch` why not call it a batch instead of `ReaderBatchWriter`? 

Also why does it return a writer instead of just being a writer?